tools/tplgtool.py

```diff
@@ -599,17 +600,136 @@ def _merge_pcm_list(self, pcm_block_list):
+                    if type(source_node["sink"]) != list:
+                        next_list = [source_node["sink"], sink_node]
+                        source_node["sink"] = next_list
+                    # if already link to a node list
+                    else:
+                        source_node["sink"].append(sink_node)
+                # not link to any node
+                else:
+                    source_node["sink"] = sink_node
+
+                if sink_node["source"] != None:
+                    if type(sink_node["source"]) != list:
+                        prev_list = [sink_node["source"], source_node]
+                        sink_node["source"] = prev_list
+                    else:
+                        sink_node["source"].append(source_node)
+                else:
+                    sink_node["source"] = source_node
+
+        link_head_list = []
+        # find head node of a graph
+        for elem in node_list:
+            if elem["source"] == None and elem["sink"] != None:
+                link_head_list.append(elem)
+        return link_head_list, node_list
+
+    # initialize node list from widget list, all the node should have its name
+    # and widget filed initialized, and other fileds are left None
+    def _init_node_list(self):
+        node_list = []
+
+        for widgets in self._tplg["widget_list"]:
+            for widget in widgets:
+                node = {"name":widget["name"], "widget":widget, "ctrl":None, "source":None, "sink":None}
+                node_list.append(node)
+        return node_list
+
+    # find node by its name from node_list, as the name of a node is not unified to
+    # widget["name"] or widget["sname"], we should check both
+    @staticmethod
+    def find_node_by_name(name, node_list):
+        if name == '': return None
+        for node in node_list:
+                if name == node["widget"]["name"] or name == node["widget"]["sname"]:
+                    return node
+        # if excution goes here, it means we didn't find the widget in the list,
+        # obviously, there is error in topology
+        print("Widget {} not exist, error in topology", name)
+        sys.exit(1)
+        return None
+
+    # find specified type of components connected to ref_node
+    @staticmethod
+    def find_connected_comp(ref_node, comp_type):
+        if ref_node == None:
+            return None
+        comp_type = comp_type.upper() # to upper case
+        comp_list = []
+        node = ref_node
+        # Currently, multiplex node is not supported, once met
+        # multiplex node, end while loop
+        while type(node) != list and node["source"] != None:
+            if node["widget"]["name"].startswith(comp_type) or \
+                node["widget"]["sname"].startswith(comp_type):
+                comp_list.append(node["widget"])
+            node = node["source"]
+
+        node = ref_node
+        while type(node) != list and node["sink"] != None:
```

Comment:
`node["sink"]` change to the list

Reply:
for most node, there is only one node connected to it. the condition a node connected to two nodes is really rare, only happens to wov pipeline. so I think we don't need to use list for every node, this add complexity for dereferencing.